### DISCUSSION TITLE
ARM build and execution?

### DISCUSSION THREAD

**Discussion by brad-4k** [valhalla/valhalla#3432] [Q&A]
Hi all,

I didn't find anything about this, if it works, any issues, etc.  With a few small changes I was able to build.  When I run valhalla_build_tiles I get a SIGSEGV pretty early on:

> $ valhalla_build_tiles -c ./conf/valhalla.json albania-latest.osm.pbf 
> 2021/11/30 03:01:46.135807 [INFO] Start stage = initialize End stage = cleanup
> 2021/11/30 03:01:46.138291 [INFO] Parsing files for ways: albania-latest.osm.pbf
> 2021/11/30 03:01:46.138415 [INFO] Parsing ways...
> Segmentation fault (core dumped)

GDB results are suspect, although maybe it's correct:

> Program received signal SIGSEGV, Segmentation fault.
> 0x0000fffff7fbe0d4 in typeinfo for boost::program_options::invalid_option_value () from /usr/lib/libboost_program_options.so.1.77.0
> (gdb) where
> #0  0x0000fffff7fbe0d4 in typeinfo for boost::program_options::invalid_option_value () from /usr/lib/libboost_program_options.so.1.77.0
> #1  0xfffd95e8d114a038 in ?? ()
> Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Any thoughts on where I should look to figure out what has gone astray?

**Comment by kevinkreiser** [comment]
yeah i would say the gdb results are very suspect given that we only interact with program options right at the beginning before we even start parsing. did you make a debug build or a release build with debug symbols?

**Comment by brad-4k** [comment]
I built with debug:

/usr/bin/valhalla_build_tiles: ELF 64-bit LSB shared object, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=969c008656f920b83a1c34c6db3c9ab45ea4a598, **_with debug_info, not stripped_**

For ARM I had to make a small change because the default char type is unsigned.  There was a bit of code that was tripping a warning, which was killing the build (-Werror).  Eventually I had to disable -Werror because of another warning (definition of minor())

The small change probably should get merged into release:

```
+++ b/valhalla/baldr/json.h
@@ -13,6 +13,7 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <limits.h>
 
 namespace valhalla {
 namespace baldr {
@@ -110,7 +111,16 @@ public:
           ostream_ << "\\t";
           break;
         default:
+         // Unsigned char (default on Arm) triggers an "always true" warning here for c >= 0 which falls out as an error.  This should fix that.
+          #ifdef __GNUC__
+          #if CHAR_MIN == SCHAR_MIN
           if (c >= 0 && c < 32) {
+         #else
+          if (c < 32) {
+          #endif
+         #else
+          if (c >= 0 && c < 32) {
+          #endif

```

**Reply by kevinkreiser** [reply]
yeah the whole platform dependent use of signed or unsigned types in std::string... we should just avoid it completely if we can. i would not make an changes here and just ignore the warnings

**Comment by brad-4k** [comment]
I'm getting closer...  through some debugger 'foo (MBP(vscode) -> SSH -> EC2 -> GDB...), I've tracked the crash to the loop in:
`OSMPBF::Parser::parse`
Not sure which call it's SIGSEGV'ing in yet because of the stack issue.

More to come but ideas still welcome.

**Comment by brad-4k** [comment]
Ok, well it makes it through about 820 iterations, crashing after the 821st.  Crashing in `parse_primitive_block(unpack_buffer, sz, interest, callback);`. 

Just for reference:
osmpbfparser.cc
```
...
  while (!file.eof()) {
    std::cout << count++ << " ";
    // grab the blob header
    bool finished = false;
    BlobHeader header = read_header(buffer, file, finished);
    // if we didnt hit the end
    if (!finished) {
      // grab the blob that goes with the blob header
      int32_t sz = read_blob(buffer, unpack_buffer, file, header);
      // if its data parse it
      if (header.type() == "OSMData") {
        parse_primitive_block(unpack_buffer, sz, interest, callback);
        // if its something other than a header
      } else if (header.type() == "OSMHeader") {
        parse_header_block(unpack_buffer, sz);
      } else {
        LOG_WARN("Unknown blob type: " + header.type());
      }
    }
  }
```

Crash area (inside loop) in `parse_primitive_block(unpack_buffer, sz, interest, callback);`:
```
...
    // do the ways
    if ((interest & WAYS) == WAYS) {
      for (const auto& way : primitive_group.ways()) {
        uint64_t node = 0;
        std::vector<uint64_t> nodes;
        nodes.reserve(way.refs_size());
        for (auto node_id_offset : way.refs()) {
          node += node_id_offset;
          // TODO: skip consecutive duplicates, make this configurable
          if (nodes.size() == 0 || node != nodes.back()) {
            nodes.push_back(node);
          }
        }
        callback.way_callback(way.id(), get_tags<Way>(way, primblock), nodes);
        if (way.has_info() && way.info().has_changeset() && (interest & CHANGESETS) == CHANGESETS) {
          callback.changeset_callback(way.info().changeset());
        }
      }
    }
...

```

**Reply by brad-4k** [reply]
Ok, I'm closer to the culprit.
pbfgraphparser.cc:1684:
```
    // Transform tags. If no results that means the way does not have tags
    // suitable for use in routing.
    Tags results =
        tags.size() == 0 ? empty_way_results_ : lua_.Transform(OSMType::kWay, osmid_, tags);
```

Causes SIGSEGV

**Reply by kevinkreiser** [reply]
well then i wonder if its just calling lua in general. can you check if you remove the ternary and just call lua regardless of tags.size(). then the question will be what is wrong with calling lua

**Reply by brad-4k** [reply]
I think it's important I list the specifics of the system I'm trying to build/run on:

Ubuntu 18.04

$ lscpu
```
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              32
On-line CPU(s) list: 0-31
Thread(s) per core:  1
Core(s) per socket:  32
Socket(s):           1
NUMA node(s):        1
Vendor ID:           ARM
Model:               1
Stepping:            r3p1
BogoMIPS:            243.75
L1d cache:           64K
L1i cache:           64K
L2 cache:            1024K
L3 cache:            32768K
NUMA node0 CPU(s):   0-31
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
```

$ lsmem
```
RANGE                                  SIZE  STATE REMOVABLE  BLOCK
0x0000000040000000-0x000000007fffffff    1G online        no      1
0x0000000400000000-0x000000433fffffff  253G online        no 16-268

Memory block size:         1G
Total online memory:     254G
Total offline memory:      0B
```

Boost built from boost_1_77_0.tar.bz2 (Ubuntu 18.04 natively installs 1.65.1)
liblua5.3-dev/bionic-updates,bionic-security,now 5.3.3-1ubuntu0.18.04.1 arm64

**Reply by brad-4k** [reply]
Another test I could perform is to build this on a Ubuntu 18.04 machine with an Intel CPU and see if I run into similar issues.  Would let me know if it's more CPU related versus LUA version (issue) related.

**Reply by brad-4k** [reply]
> well then i wonder if its just calling lua in general. can you check if you remove the ternary and just call lua regardless of tags.size(). then the question will be what is wrong with calling lua

It still SIGSEGV's at the lua call but I think memory is just corrupted before that call. I get through more iterations, at times, while stepping through with the debugger.

**Reply by kevinkreiser** [reply]
FWIW we stopped supporting 18.04 a while ago. thats not to say it wont work, it could but we no longer test on that platform.

**Reply by brad-4k** [reply]
> FWIW we stopped supporting 18.04 a while ago. thats not to say it wont work, it could but we no longer test on that platform.

Understood.  Maybe, still worth a shot.  During build, I encountered this error:
pbfgraphparser.cc:1423:20: error: ‘class boost::optional<std::unordered_map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> > >’ has no member named ‘has_value’; did you mean ‘value’?

... possible culprit? And now for the life of me, I can't remember how I got rid of it.

**Reply by kevinkreiser** [reply]
for has_value you can just cast to bool, im pretty sure the boolean operator existed in all versions of boost optional

**Reply by brad-4k** [reply]
> for has_value you can just cast to bool, im pretty sure the boolean operator existed in all versions of boost optional

Is the has_value() method in boost::optional (or parent) or is it trying to come from the std::any class?  Not having much luck.  I am using boost 1.77.

**Reply by brad-4k** [reply]
Ok, I'm thick headed.  This change worked:

`if (!((bool)results) || (found = results->find("amenity")) == results->end() ||`

**Reply by nilsnolde** [reply]
see #3425 which is also fixing that. don't think you need to explictly cast here, !results should work

**Reply by brad-4k** [reply]
> see #3425 which is also fixing that. don't think you need to explictly cast here, !results should work

Yep, !results worked.  I was just being pedantic.

**Comment by brad-4k** [comment]
Ok, just to remove the CPU differences (ARM/intel) from the equation and focus on the Ubuntu version differences (18.04/20.0?), I built on Intel Ubuntu 18.04 and tried to run.  Still get a crash.  Building and running with Debug enabled it hangs.  Just trying to be thorough.
